janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java

```diff
@@ -416,17 +452,107 @@ public void bulkRequest(List<ElasticSearchMutation> requests, String ingestPipel
             APPEND_OP.apply(builder).append("refresh=").append(bulkRefresh);
         }
         builder.insert(0, REQUEST_SEPARATOR + "_bulk");
+        return Pair.with(builder.toString(), outputStream.toByteArray());
+    }
+
+    private List<Triplet<Object, Integer, RequestBytes>> pairErrorsWithSubmittedMutation(
+        //Bulk API is documented to return bulk item responses in the same order of submission
+        //https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html#bulk-api-response-body
+        //As such we only need to retry elements that failed
+        final List<Map<String, RestBulkResponse.RestBulkItemResponse>> bulkResponseItems,
+        final List<RequestBytes> submittedBulkRequestItems) {
+        final List<Triplet<Object, Integer, RequestBytes>> errors = new ArrayList<>(bulkResponseItems.size());
+        for (int itemIndex = 0; itemIndex < bulkResponseItems.size(); itemIndex++) {
+            Collection<RestBulkResponse.RestBulkItemResponse> bulkResponseItem = bulkResponseItems.get(itemIndex).values();
+            if (bulkResponseItem.size() > 1) {
+                throw new IllegalStateException("There should only be a single item per bulk reponse item entry");
+            }
+            RestBulkResponse.RestBulkItemResponse item = bulkResponseItem.iterator().next();
+            if (item.getError() != null && item.getStatus() != 404) {
```

Comment:
(nit) can we avoid explicitly using `404`? ES library should have a constant variable for 404.

Reply:
No objection from me it's just there because [it's what was there previously](https://github.com/JanusGraph/janusgraph/blob/e30b4d6ed84e243eddbc726e4304edd1091c3eb3/janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java#L425).

Happy to change it. That's being compared against something from the Apache Http library, and it does provide status code constants like `HttpStatus.SC_NOT_FOUND`.

Switched in 09891809e

Reply:
Oh I didn’t notice that. Thx for polishing it!